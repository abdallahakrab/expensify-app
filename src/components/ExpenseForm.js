import React from 'react';
import moment from 'moment';
import {SingleDatePicker} from 'react-dates';
import 'react-dates/lib/css/_datepicker.css';

const now = moment() ; // we get instance of moment back , not passing to it anything will return current point of time 

console.log(now.format());  // to get date out we use .format method without args year-month-day-time see docs

export default class ExpenseForm extends React.Component {

    state = {
        description: '',
        note: '',
        amount: '',
        createdAt: moment(),
        focused: false,
        error: ''
    };

     onDescriptionChange = (e) => {
        const description = e.target.value;
        this.setState(()=>({description}));
    };

    onNoteChange = (e) => {
        const note = e.target.value;
        this.setState(() => ({note}));
    };

    onAmountChange = (e) => {
        const amount = e.target.value;
        if(!amount ||amount.match(/^\d{1,}(\.\d{0,2})?$/)) {
        this.setState(() => ({amount}));
        };
    };

    onDateChange = (createdAt) => {
        if(createdAt)
        this.setState(() => ({createdAt }));
    };

    onFocusChange = ({focused}) =>{
         this.setState(() => ({focused}));
    };

    onSubmit = (e) => { 
        e.preventDefault();

        if(!this.state.description || !this.state.amount) {
            this.setState(()=>({
                error: 'please provide description and amount to continue'
            })) ;
        } else {
            this.setState(()=>({error: ''}));
        }

    };


    render(){
        return (
            <div>
            {this.state.error && <p>{this.state.error}</p>}
             <h1>Add Expense</h1>
             <form onSubmit={this.onSubmit}>
              <input
               type="text"
               placeholder="Description"
               autoFocus
               value={this.state.description}
               onChange={this.onDescriptionChange}
              />
              <input 
               type="text"
               placeholder="Amount"
               value={this.state.amount}
               onChange={this.onAmountChange}
              />

              <SingleDatePicker
               date={this.state.createdAt}
               onDateChange={this.onDateChange}
               focused={this.state.focused}
               onFocusChange={this.onFocusChange}
               numberOfMonths={1}
               isOutsideRange={() => false}
              />
              <textarea
              placeholder="add a note for your expense (optional)"
              value={this.state.note}
              onChange={this.onNoteChange}
              >
              </textarea>
              <button>Add Expense</button>



             </form>
             
             
            </div>)
    };
};



