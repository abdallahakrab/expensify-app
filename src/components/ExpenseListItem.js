import React from 'react';
import {connect} from 'react-redux';

const ExpenseListItem = (props) => (
    <div>
    <p>{props.expenseDesc} {props.expenseAmount} {props.expenseCreatedAt}</p>
    </div>
);

    // const mapStateToProps = (state) => {


    //     return {
    //         description: state.expenses.description,
    //         amount: state.expenses.amount ,
    //         createdAt: state.expenses.createdAt 
            
    //     };
    // };

    
 
export default ExpenseListItem;


