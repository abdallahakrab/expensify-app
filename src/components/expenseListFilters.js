import React from 'react';
import {connect} from 'react-redux';
import {setupTextFilter, sortByAmount, sortByDate, setStartDate, setEndDate} from '../actions/filters';
import {DateRangePicker} from 'react-dates'


export class ExpenseListFilters extends React.Component {
    
    state = {
        focusedInput: null
    };

    onDatesChange = ({startDate,endDate}) => {
        this.props.setStartDate(startDate);
        this.props.setEndDate(endDate);

    };

    onFocusChange = (calendarFocused) =>{
        this.setState(()=>({calendarFocused }))
    };

    onTextChange = (e) => {
        this.props.setupTextFilter(e.target.value);
    };

    onSelectChange = (e) => {
        (e.target.value === 'date' ? this.props.sortByDate() : this.props.sortByAmount() );
        
    };
   
    render () {
       return  (
            <div>
             <input type="text" value={this.props.filters.text} onChange={this.onTextChange}/>
             <select value={this.props.filters.sortBy} onChange={this.onSelectChange}>
                 
             <option  disabled>sort</option>
              <option value="date">Date</option>
              <option value="amount">Amount</option>
        
             </select>

             <DateRangePicker
              startDate={this.props.filters.startDate}
              endDate={this.props.filters.endDate}
              onDatesChange={this.onDatesChange}
              focusedInput={this.state.focusedInput}
              onFocusChange={this.onFocusChange}
              numberOfMonths={1}
              isOutsideRange={()=>false}
              showClearDates={true}

               />
            
            </div>
        );
             }
};





const mapStateToProps = (state) => {
    return {
        filters: state.filters
    };
}; 

const mapDispatchToProps = (dispatch) => {
    return {
        setStartDate: (startDate) => { dispatch(setStartDate(startDate)) },
        setEndDate: (endDate) => {dispatch(setEndDate(endDate))},
        setupTextFilter: (value) => {dispatch(setupTextFilter(value))},
        sortByDate: () => {dispatch(sortByDate())},
        sortByAmount: () => {dispatch(sortByAmount())}

    };
};


export default connect(mapStateToProps, mapDispatchToProps)(ExpenseListFilters);