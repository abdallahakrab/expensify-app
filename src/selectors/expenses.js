

const getVisibleExpenses = (expenses,{text, sortBy, startDate, endDate}) => {
    return expenses.filter((expense) => {
        const startDateMatch = typeof startDate !== 'number' || expense.createdAt >= startDate;
        const endDateMatch = typeof endDate !== 'number' || expense.createdAt <= endDate;
        const lowerCaseText = text.toLowerCase();
        const lowerCaseDescription = expense.description.toLowerCase();
        const textMatch= typeof text !== 'string' || lowerCaseDescription.includes(lowerCaseText);

        return startDateMatch && endDateMatch && textMatch;
    }).sort((a,b) =>{
        if(sortBy === 'date')
        return ( a.createdAt > b.createdAt ? -1:1 );

        return (a.amount > b.amount ? -1:1);
    });
};

export default getVisibleExpenses;