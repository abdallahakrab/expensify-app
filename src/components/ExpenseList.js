import React from 'react';
import {connect} from 'react-redux';
import ExpenseListItem from './ExpenseListItem';
import getVisibleExpenses from '../selectors/expenses';


const ExpenseList = (props) => (
    <div>
     <h1>Expense List</h1>
    {props.expenses.map((expense) =>{
        return <ExpenseListItem
         key={expense.id} 
         expenseDesc={expense.description}
         expenseAmount={expense.amount}
         expenseCreatedAt={expense.createdAt}  /> ;
    })}
    
    </div>
);

const mapStateToProps = state => ({
    expenses: getVisibleExpenses(state.expenses, state.filters) 
});

export default connect(mapStateToProps)(ExpenseList); //here we get something back: not the hoc but the function  

