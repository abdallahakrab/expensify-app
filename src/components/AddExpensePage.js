import React from 'react';
import ExpenseForm from './ExpenseForm';
import {addExpense} from '../actions/expenses';
import { connect } from 'react-redux';

const AddExpensePage = (props) => (
    <div>
    <h3>Add Expense</h3>
     <ExpenseForm onSubmit={(expense) => {    //function defined here.
       props.dispatch(addExpense(expense));
       props.history.push('/');
     }} />
    </div>
);

export default connect()(AddExpensePage);