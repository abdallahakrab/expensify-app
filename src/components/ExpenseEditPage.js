import React from 'react';
import ExpenseForm from './ExpenseForm';
import {connect} from 'react-redux';
import {editExpense, startRemoveExpense} from '../actions/expenses';


export class ExpenseEditPage extends React.Component {

    onClick = () => { 
        this.props.startRemoveExpense({ id: this.props.Expense.id } );
        this.props.history.push('/');

    };
 
    onSubmit = (expense) => {
        this.props.editExpense(this.props.Expense.id, expense);
        this.props.history.push('/');
    };

    render() {
        return (
                <div>
                 <h3>Edit Expense</h3>
                 <ExpenseForm
                  expense={this.props.Expense}
                  onSubmit={ this.onSubmit }
                 />
                 <button onClick={ this.onClick }
                 >
                 Remove Item
                 </button>
                </div>

              );
    };

};

const mapStateToProps = (state, props) => { 
    return {Expense: state.expenses.find( (expense) => props.match.params.id === expense.id )};
};

const mapDispatchToProps = (dispatch, props) => {
    return {
        editExpense: (id,expense) => { dispatch(editExpense(id,expense)) },
        startRemoveExpense: (data) => {dispatch(startRemoveExpense(data))}
    };
};
    



export default connect(mapStateToProps,mapDispatchToProps)(ExpenseEditPage);