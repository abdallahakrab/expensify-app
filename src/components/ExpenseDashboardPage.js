import React from 'react';
import ExpenseList from './ExpenseList';
import ExpenseListFilter from './expenseListFilters';
import ExpenseSummary from './ExpensesSummary';

const ExpenseDashboardPage = () => (
    <div>
    <ExpenseListFilter/>
    <ExpenseSummary />
     <ExpenseList />
    </div>
);

export default ExpenseDashboardPage;