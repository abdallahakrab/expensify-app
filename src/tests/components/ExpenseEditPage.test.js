import React from 'react';
import { shallow } from 'enzyme';
import { ExpenseEditPage } from '../../components/ExpenseEditPage';
import expenses from '../fixtures/expenses';

let wrapper, editExpense, removeExpense, history;

beforeEach(() => {
    editExpense = jest.fn();
    removeExpense = jest.fn();
    history = {push: jest.fn()};
    wrapper = shallow(
        <ExpenseEditPage 
         Expense={expenses[1]} 
         removeExpense={removeExpense} 
         editExpense={editExpense} 
         history={history} 
        />
    );


});

test('should render ExpenseEditPage correctly', () => {
    expect(wrapper).toMatchSnapshot();
});

test('should handle edit expense', () => {
    wrapper.find('ExpenseForm').prop('onSubmit')(expenses[1]);
    expect(history.push).toHaveBeenLastCalledWith('/');
    expect(editExpense).toHaveBeenLastCalledWith(expenses[1].id,expenses[1]);

});

test('should handle removeExpense', () => {
    wrapper.find('button').simulate('click');
    expect(history.push).toHaveBeenLastCalledWith('/');
    expect(removeExpense).toHaveBeenLastCalledWith({id: expenses[1].id})
})