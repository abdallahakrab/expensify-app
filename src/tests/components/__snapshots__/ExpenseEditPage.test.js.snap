// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render ExpenseEditPage correctly 1`] = `
<div>
  <h3>
    Edit Expense
  </h3>
  <ExpenseForm
    expense={
      Object {
        "amount": 75000,
        "createdAt": 345600000,
        "description": "rent",
        "id": "2",
        "note": "",
      }
    }
    onSubmit={[Function]}
  />
  <button
    onClick={[Function]}
  >
    Remove Item
  </button>
</div>
`;
